src/OpenIddict.Server.Quartz/OpenIddictServerQuartzJob.cs

```diff
@@ -0,0 +1,154 @@
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="OpenIddictServerQuartzJob"/> class.
+        /// </summary>
+        public OpenIddictServerQuartzJob() => throw new InvalidOperationException(SR.GetResourceString(SR.ID1081));
+
+        /// <summary>
+        /// Creates a new instance of the <see cref="OpenIddictServerQuartzJob"/> class.
+        /// </summary>
+        /// <param name="logger">The logger.</param>
+        /// <param name="options">The OpenIddict server Quartz.NET options.</param>
+        /// <param name="provider">The service provider.</param>
+        public OpenIddictServerQuartzJob(
+            ILogger<OpenIddictServerQuartzJob> logger,
+            IOptionsMonitor<OpenIddictServerQuartzOptions> options,
+            IServiceProvider provider)
+        {
+            _logger = logger;
+            _options = options;
+            _provider = provider;
+        }
+
+        /// <summary>
+        /// Gets the default identity assigned to this job.
+        /// </summary>
+        public static JobKey Identity { get; } = new JobKey(
+            name: typeof(OpenIddictServerQuartzJob).Name,
+            group: typeof(OpenIddictServerQuartzJob).Assembly.GetName().Name!);
+
+        /// <inheritdoc/>
+        public async Task Execute(IJobExecutionContext context)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            // TODO: determine whether errors should be swallowed or should be re-thrown to be logged by Quartz.NET.
+
+            // Note: this job is registered as a transient service. As such, it cannot directly depend on scoped services
+            // like the core managers. To work around this limitation, a scope is manually created for each invocation.
+            var scope = _provider.CreateScope();
+
+            try
+            {
+                // Note: this background task is responsible of automatically removing orphaned tokens/authorizations
+                // (i.e tokens that are no longer valid and ad-hoc authorizations that have no valid tokens associated).
+                // Since ad-hoc authorizations and their associated tokens are removed as part of the same operation
+                // when they no longer have any token attached, it's more efficient to remove the authorizations first.
+
+                // Note: the authorization/token managers MUST be resolved from the scoped provider
+                // as they depend on scoped stores that should be disposed as soon as possible.
+
+                if (!_options.CurrentValue.DisableAuthorizationsPruning)
+                {
+                    var manager = scope.ServiceProvider.GetService<IOpenIddictAuthorizationManager>();
+                    if (manager == null)
+                    {
+                        throw new InvalidOperationException(SR.GetResourceString(SR.ID1277));
```

Comment:
Ideally you should only throw `JobExecutionException` from Execute, any other exception signals scheduler error which "a little bit more severe". `JobExecutionException` also allows you to instruct scheduler like "remove triggers, this won't ever work" or "re-fire immedialtely, we can do this!"

Reply:
Wooot, that's indeed much better, thanks for the info!

> "re-fire immedialtely, we can do this!"

It would be a nice option for transient option. Is there a limit to prevent a job from returning `JobExecutionException`s with `RefireImmediately` set to `true` indefinitely or is it something that should be done inside the job itself?

Reply:
If you signal the `RefireImmediately = true` you [start looping inside the job invocation logic](https://github.com/quartznet/quartznet/blob/9dfe5ee5ee93a53cea87b9950234c0012afac9f1/src/Quartz/Core/JobRunShell.cs#L255-L271). This is basically the same thing as if you would do the loop with try-catch inside your job trying to get successful execution. So if you consider exception transient, you should signal the flag. This however means that you job execution time could vary from normal 2 seconds to 5 minutes. If you job would run every 1 minute there would be misfires (concurrent execution disallowed) and you would need to have misfire policy in place, like Ignore, continue with next scheduled time when it comes.

You will also get the misfire count from job execution context in your job if you want to abort trying after some trials.

Reply:
> This however means that you job execution time could vary from normal 2 seconds to 5 minutes. If you job would run every 1 minute there would be misfires (concurrent execution disallowed) and you would need to have misfire policy in place, like Ignore, continue with next scheduled time when it comes.

Thanks for those precious details!

> You will also get the misfire count from job execution context in your job if you want to abort trying after some trials.

Perfect. I'll update the job to catch those transient exceptions and re-throw `JobExecutionException`s (and check the `RefireCount` property to ensure it doesn't cause endless loops 😃)